pallets/elections/src/impls.rs

```diff
@@ -0,0 +1,398 @@
+use crate::{Config, ErasReserved, MembersPerSession, Pallet, SessionValidatorBlockCount};
+use frame_election_provider_support::sp_arithmetic::Perquintill;
+use frame_support::{pallet_prelude::Get, traits::Currency};
+use sp_staking::{EraIndex, SessionIndex};
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+const MAX_REWARD: u32 = u32::MAX;
+const LENIENT_THRESHOLD: Perquintill = Perquintill::from_percent(90);
+
+fn calculate_adjusted_session_points(
+    sessions_per_era: EraIndex,
+    blocks_to_produce_per_session: u32,
+    blocks_created: u32,
+    total_exposure_in_tokens: u32,
+) -> u32 {
+    let performance =
+        Perquintill::from_rational(blocks_created as u64, blocks_to_produce_per_session as u64);
+
+    // when produced between 90% to 100% expected blocks get 100% possible reward for session
+    if performance >= LENIENT_THRESHOLD && blocks_to_produce_per_session >= blocks_created {
+        return (Perquintill::from_rational(1, sessions_per_era as u64)
+            * total_exposure_in_tokens as u64) as u32;
+    }
+
+    (Perquintill::from_rational(
+        blocks_created as u64,
+        (blocks_to_produce_per_session * sessions_per_era) as u64,
+    ) * total_exposure_in_tokens as u64) as u32
```

Comment:
There is this question of whether this should be, like that, or it should be further divided by `LENIENT_THRESHOLD` so that the reward function is continuous. I don't have a strong preference here.

Reply:
me neither, can we leave it like it is for now?